File: src/main/java/io/github/liquidTruffle/parser/ast/nodes/VariableNode.java
package io.github.liquidTruffle.parser.ast.nodes;

import io.github.liquidTruffle.parser.ast.AstNode;
import io.github.liquidTruffle.runtime.LiquidRuntime;
import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.nodes.NodeInfo;
import java.util.List;
import java.util.ArrayList;

@NodeInfo(description = "Represents variable references with optional filters in Liquid templates")
public class VariableNode extends AstNode {
    private final String name;
    private final List<FilterSpec> filters;

    public VariableNode(String name, List<FilterSpec> filters) {
        this.name = name;
        this.filters = filters;
    }

    @Override
    public Object executeGeneric(VirtualFrame frame) {
        Object current = LiquidRuntime.getVariable(frame, name);
        for (FilterSpec filter : filters) {
            List<Object> evaluatedArgs = new ArrayList<>();
            for (AstNode arg : filter.getArgs()) {
                evaluatedArgs.add(arg.executeGeneric(frame));
            }
            current = LiquidRuntime.applyFilter(frame, filter.getName(), current, evaluatedArgs);
        }
        return current;
    }
    
    // Expose fields for testing purposes
    public String getVariableName() {
        return this.name;
    }
    
    public List<FilterSpec> getFilterSpecs() {
        return this.filters;
    }

    public static class FilterSpec {
        private final String name;
        private final List<AstNode> args;

        public FilterSpec(String name, List<AstNode> args) {
            this.name = name;
            this.args = args != null ? args : new ArrayList<>();
        }

        public String getName() {
            return name;
        }

        public List<AstNode> getArgs() {
            return args;
        }
    }
}

File: src/main/kotlin/io/github/liquidTruffle/ast/nodes/VariableNode.kt
package io.github.liquidTruffle.ast.nodes

import io.github.liquidTruffle.ast.AstNode
import io.github.liquidTruffle.runtime.LiquidRuntime
import com.oracle.truffle.api.frame.VirtualFrame

class VariableNode(
	private val name: String,
	private val filters: List<FilterSpec>
) : AstNode() {
	
	override fun executeGeneric(frame: VirtualFrame): Any? {
		var current = LiquidRuntime.getVariable(frame, name)
		for (filter in filters) {
			current = LiquidRuntime.applyFilter(frame, filter.name, current, filter.args)
		}
		return current
	}
	
	// Expose fields for testing purposes
	val variableName: String get() = this.name
	val filterSpecs: List<FilterSpec> get() = this.filters

	data class FilterSpec(
		val name: String,
		val args: List<Any?> = emptyList()
	)
}
